DS-3517 Allow improved handling of CMYK PDFs

alanorth opened the pull request:
https://jira.duraspace.org/browse/DS-3517
https://jira.duraspace.org/browse/DS-2834

The ImageMagick-based thumbnailing in DSpace 5+ should explicitly generate JPGs in the sRGB color space, which is the default color space for the web. In the case where an input file is in another color space, like CMYK, ImageMagick happily generates a CMYK JPG, which leads to distorted colors.

It should be noted that this is a merely a regression in the DSpace 5+ thumbnail code, as the XPDF-based thumbnailing in DSpace 3 and 4 was correctly producing sRGB images.

alanorth commented:
Sorry, I'm not sure if this is better to apply against master and then back port to `dspace-5_x` and `dspace-6_x`?

tdonohue commented:
@alanorth : with smaller PRs like this, it doesn't much matter which branch you create the PR against. We can always cherry-pick its commit(s) to other appropriate locations (in this case `dspace-6_x` and `master`).

That said, from the ticket comments, it sounds like you are still working on this code. I'm flagging this as a work in progress for now. Let us know when you feel this is ready for review.

alanorth commented:
Great. Yes, I'm still working on it. Comments in Jira [DS-3517](https://jira.duraspace.org/browse/DS-3517).

alanorth committed:
DS-3517 Allow improved handling of CMYK PDFs

Allow ImageMagick to generate thumbnails with more accurate colors
for PDFs using the CMYK color system. This adds two options to the
dspace.cfg where the user can optionally specify paths to CMYK and
RGB color profiles if they are available on their system (they are
provided by Ghostscript 9.x).

Uses im4java's Info class to determine the color system being used
by the PDF.

See: http://im4java.sourceforge.net/docs/dev-guide.html

alanorth commented:
Reworked this, now adds two new config options and basic wiring to detect / handle CMYK PDFs properly using `org.im4java.core.Info`. I need feedback on how to properly access variables from the class' `static` scope in the `getImageFile()` function, though (see comments in the code)!

alanorth commented:
@tdonohue this is functionally complete but just needs advice on reading config variables. Thanks.

terrywbrady commented on dspace-api/src/main/java/org/dspace/app/mediafilter/ImageMagickThumbnailFilter.java:
```diff
@@ -132,12 +136,22 @@ public static File getImageFile(File f, int page) throws IOException, Interrupte
     	ConvertCmd cmd = new ConvertCmd();
 		IMOperation op = new IMOperation();
+		Info imageInfo = new Info(f.getAbsolutePath(),true);
+		// I'd rather set these in the class but can't figure out how to access them here if I do!!!
```
Since the values of the config settings contain a path to the ghostscript install, I think you will need to continue setting this inside of the config file.

terrywbrady replied:
Perhaps I understand the question after taking a closer look.  I recommend that you move the declaration for cmyk_profile  and srgb_profile before the static { ...} block and declare them as static.  Initialize them inside the static {} block.  You can then reference the initialized values in getImageFile()

alanorth commented:
Cleaned up the style with @terrywbrady's suggestions about static variables declaration/initialization, and now ready for @tdonohue.

terrywbrady commented:
Thanks @alanorth, I removed the work in progress to alert others that this is open for review.  The code looks good.  I will run a quick test as well.

alanorth commented:
I updated the commit message to give more information about the extra im4java import that I added.

tdonohue reviewed (approved):
👍 Code looks reasonable to me. Needs a test before merger. 

After merger, this should also be applied (e.g. cherry-picked) to `dspace-6_x` and `master` branches.  Also will require documentation updates (obviously)

alanorth commented:
Great, @tdonohue. If you give me write access to the wiki I can add the documentation, otherwise I can just write the blurbs to be added.

tdonohue commented:
@alanorth : you should now have edit access to all the official documentation (all DSDOC* spaces in wiki). I'd recommend still waiting until this gets fully tested to add the edits, but you are welcome to begin drafting them.

terrywbrady commented:
@alanorth , sorry for the delay in testing. 

I have installed this PR on my server.

When I attempted to configure this feature, I see that I am running ghostscript 8.70 on my test box. I am unable to find the icc files within my ghostscript distribution.

Is there a dependency on specific versions of ghostscript, or is it likely that I will need to hunt for these files?

alanorth commented:
Thanks, @tdonohue—I assume it's just the dspace.cfg section and perhaps the filter-media section that need a blurb? @terrywbrady yes, in my test it looks like the ICC profiles are only on Ghostscript 9 and up. I suspect you're on CentOS 6? I don't know what happens if you download the profiles and try to use them on older Ghostscript, but I should note this version dependency in the dspace.cfg blurb etc.

terrywbrady commented:
It would be good to find a tester with GS 9+ installed.  It should be a very quick test for someone to run.

terrywbrady reviewed (commented):
Code looks good.  I am unable to test.  I recommend finding a tester with GS 9+ installed.

alanorth commented:
@terrywbrady I amended the commit message and `dspace.cfg` note to mention Ghostscript 9.x requirement. I will post on the mailing list to ask if someone can test this.

pnbecker commented on dspace-api/src/main/java/org/dspace/app/mediafilter/ImageMagickThumbnailFilter.java:
```diff
@@ -132,12 +137,19 @@ public static File getImageFile(File f, int page) throws IOException, Interrupte
+		// PDFs using the CMYK color system can be handled specially if profiles are defined
+		if (imageClass.contains("CMYK") && cmyk_profile != null && srgb_profile != null) {
+			op.profile(cmyk_profile);
+			op.profile(srgb_profile);
```
In imagemagick's documentation is written: "If there is no embedded profile then the first "-profile" is the input profile. A second "-profile" then defines the output profile. If there is an embedded profile then a single of "-profile" operator will immediately define the output profile" (see http://www.imagemagick.org/Usage/formats/#color_profile_changing). Shouldn't we check if a color profile is embedded within the PDF before assigning two profiles? While the output profile should be always fine, especially the input profile should match the source's profile as close as possible.

alanorth replied:
I suppose I could add some more conditional logic to test if there was an embedded profile, though I'm not sure how common PDFs with embedded profiles would be. Anecdotally, on our repository I re-processed all our PDFs bitstreams and noted that ~3,000 items out of 15,000 were using the CMYK color system, which is a pretty big win if you ask me!

I will try to find (or create) some PDFs with embedded profiles to do some experiments with this this week.

pnbecker reviewed (changes requested):
I love the general approach here. But we should check if there are profiles embedded into the source before applying the default cmyk profile. @alanorth Is there a way to check that?

pnbecker commented:
@alanorth thanks for your prompt reaction.
Your PR is a big win, no doubt about that. The question is how many of the 3.000 pdfs that uses CMYK do contain an ICC profile. As CMYK is used especially in the printing business I would expect PDFs using CMYK to have more often an embedded profile than pdfs that use RGB.

pnbecker commented:
I found a PDF and was able to test this PR. It works as expected. I will retest and review it as soon as the handling of files with embedded icc profiles is enhanced.

pnbecker commented:
@alanorth With the code described [here](https://stackoverflow.com/questions/40521914/how-do-you-extract-color-profiles-from-a-pdf-file-using-pdfbox-or-other-open-so#40588767) I was able to extract an ISO Coated v2 (ECI) icc profile from the pdf here: http://dx.doi.org/10.14279/depositonce-5383.
DSpace 6.1 shall be released asap. I would love to merge this PR by the mid of next week. So please let me know if you see any chance to work on this or if we need to look out if someone else may help.

alanorth commented:
That's great news. Yes, I'm going to be very busy in the next week so I'd
appreciate someone to help finish the implementation in this PR.

On Fri, Jun 2, 2017, 14:37 Pascal-Nicolas Becker <notifications@github.com>
wrote:

> @alanorth <https://github.com/alanorth> With the code described here
> <https://stackoverflow.com/questions/40521914/how-do-you-extract-color-profiles-from-a-pdf-file-using-pdfbox-or-other-open-so#40588767>
> I was able to extract an ISO Coated v2 (ECI) icc profile from the pdf here:
> http://dx.doi.org/10.14279/depositonce-5383.
> DSpace 6.1 shall be released asap. I would love to merge this PR by the
> mid of next week. So please let me know if you see any chance to work on
> this or if we need to look out if someone else may help.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DSpace/DSpace/pull/1669#issuecomment-305762540>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALtCnC9mgdlxYrCIe9bnEc4VpqinCY4ks5r__PhgaJpZM4MSFiG>
> .
>
-- 
Alan Orth
alan.orth@gmail.com
https://picturingjordan.com
https://englishbulgaria.net
https://mjanja.ch

tdonohue commented:
@pnbecker : I'm unclear of the latest status of this PR. Are you satisfied with the PR, or are further changes needed?  Looks like your review is blocking merging this PR

pnbecker commented:
I'd say lets merge it as it is. I create a ticket tomorrow about what should be enhanced later. This PR in is current form is already a big step in the right direction.

pnbecker commented:
I created a ticket regarding the outstanding possible improvement: https://jira.duraspace.org/browse/DS-3614 and set it to "Needs Volunteer".
